expr/service.go

```diff
@@ -133,7 +123,34 @@ func (e *ErrorExpr) Validate() error {
 // Note: this may produce a user type with an attribute that is not an object!
 func (e *ErrorExpr) Finalize() {
 	att := e.AttributeExpr
-	if _, ok := att.Type.(UserType); !ok {
+	switch dt := att.Type.(type) {
+	case UserType:
+		if dt != ErrorResult {
+			// If this type contains an attribute with "struct:error:name" meta
+			// then no need to do anything.
+			var found bool
+			for _, nat := range *AsObject(dt) {
+				if _, ok := nat.Attribute.Meta["struct:error:name"]; ok {
+					found = true
+					break
+				}
+			}
+			if found {
+				return
+			}
+
+			// This type does not have an attribute with "struct:error:name" meta.
+			// It means the type is used by at most one error (otherwise validations
+			// would have failed). Dup type and add "struct:error:name" meta to
+			// indicate that the type is an error type.
+			dt = Dup(att.Type).(UserType)
```

Comment:
Not sure we need to dup the type? duping is expensive and as you know can have interesting side effects (e.g. generated types etc.)